bcp47/library/Data/BCP47/Internal/Region.hs

```diff
@@ -1,37 +1,48 @@
 module Data.BCP47.Internal.Region
-  ( CountryCode
+  ( Country
   , regionToText
   , regionFromText
   , regionP
   )
 where
 
+import Control.Applicative ((<|>))
+import Country (Country, alphaTwoUpper, decodeAlphaTwo, decodeNumeric)
 import Data.BCP47.Internal.Parser (complete)
 import Data.Bifunctor (first)
-import Data.ISO3166_CountryCodes (CountryCode)
 import Data.Text (Text, pack)
 import Data.Void (Void)
-import Text.Megaparsec (Parsec, count, parse)
-import Text.Megaparsec.Char (upperChar)
+import Text.Megaparsec (Parsec, count, parse, try)
+import Text.Megaparsec.Char (digitChar, upperChar)
 import Text.Megaparsec.Error (errorBundlePretty)
 import Text.Read (readEither)
 
-regionToText :: CountryCode -> Text
-regionToText = pack . show
+regionToText :: Country -> Text
+regionToText = alphaTwoUpper
 
 -- | Parse a region subtag from 'Text'
-regionFromText :: Text -> Either Text CountryCode
+regionFromText :: Text -> Either Text Country
 regionFromText =
   first (pack . errorBundlePretty) . parse regionP "regionFromText"
 
 -- | BCP-47 region parser
 --
--- This only implements the ISO portion of the grammar.
---
 -- @@
 -- region        = 2ALPHA              ; ISO 3166-1 code
 --               / 3DIGIT              ; UN M.49 code
 -- @@
 --
-regionP :: Parsec Void Text CountryCode
-regionP = complete $ either fail pure . readEither =<< count 2 upperChar
+regionP :: Parsec Void Text Country
+regionP = complete $ try alpha2 <|> num3
```

Comment:
Minor, but I think when you do `p1 <|> p2` it can be frustrating for an end-user who supplies data that is intended for `p1` but is invalid, because the error will come from `p2`. IIRC, `Parsec` has an operator for attaching our own message, which might be helpful here?

```hs
alpha2 <|> num3 <?> "Must be 2 or 3 character country code"
```

Reply:
:+1: